momaland/envs/breakthrough/breakthrough.py

```diff
@@ -0,0 +1,355 @@
+                        legal_moves.add(self._move_to_int(x, y, "RIGHT"))
+                if x > 0:
+                    if (
+                        self._get_square(x - 1, y + move_direction) == 0
+                        or self._get_square(x + 1, y + move_direction) == opp_piece
+                    ):  # move diagonally to the left possible
+                        legal_moves.add(self._move_to_int(x, y, "LEFT"))
+        return legal_moves
+
+    def _get_square(self, x, y):
+        """Returns the piece at the given coordinates, provided the coordinates are legal. Otherwise returns MOBreakthrough.OFF_BOARD."""
+        if x < 0 or x > self.board_width - 1 or y < 0 or y > self.board_height - 1:
+            return MOBreakthrough.OFF_BOARD
+        return self.board[x][y]
+
+    @override
+    def step(self, action):
+        if self.truncations[self.agent_selection] or self.terminations[self.agent_selection]:
+            return self._was_dead_step(action)
+
+        # assert valid move
+        assert action in self.legal_moves, "played illegal move."
+
+        # make the move
+        x, y, direction = self._int_to_move(action)
+        agent = self.agent_selection
+        agent_index = self.possible_agents.index(agent)
+        next_agent = self.possible_agents[(agent_index + 1) % 2]
+        agent_piece = agent_index + 1
+        move_direction = 1 if agent_piece == 1 else -1
+        self.board[x][y] = 0
+        capture = False
+        if self.board[x + MOBreakthrough.ANGLES.index(direction) - 1][y + move_direction] != 0:
+            capture = True
+        self.board[x + MOBreakthrough.ANGLES.index(direction) - 1][y + move_direction] = agent_piece
+        self.move_count += 1
+
+        # handle the rewards
+        self.rewards = {i: np.array([0] * self.num_objectives, dtype=np.float32) for i in self.agents}
+        if capture:
+            if self.num_objectives > 2:
+                self.rewards[agent][2] = 1 / (self.board_width * 2)
+            if self.num_objectives > 3:
+                self.rewards[next_agent][3] = -1 / (self.board_width * 2)
+        if self.check_for_winner():
+            self.rewards[agent][0] = 1
+            self.rewards[next_agent][0] = -1
+            if self.num_objectives > 1:
+                self.rewards[agent][1] = 1 - (self.move_count / self.max_turns)
+                self.rewards[next_agent][1] = -(1 - (self.move_count / self.max_turns))
+            self.terminations = {i: True for i in self.agents}
+        self._cumulative_rewards[agent] = np.array([0] * self.num_objectives, dtype=np.float32)
+        self._accumulate_rewards()
+
+        # select the next agent
+        self.agent_selection = self._agent_selector.next()
+        self.legal_moves = self._legal_moves()
+        if self.render_mode == "human":
+            self.render()
+
+    @override
+    def reset(self, seed=None, options=None):
+        if seed is not None:
+            np.random.seed(seed)

```

Comment:
I think the PZ convention is to store `self.seed` and `self.np_random` from that call now. And use the `self.np_random` for rng stuff

Reply:
np.random generators seem to not actually have all their state defined by the seed alone, which is why I store/restore their entire state in samegame.py. What do you think about that solution?

Reply:
So the only stochastic environment is samegame, right?

Reply:
Right